File: ruby/page_objects/pages/login_page.rb
class LoginPage < SitePrism::Page
  set_url 'practice-test-login'
  load_validation { has_user_name? }

  element :user_name, 'input[name="username"]'
  element :password, 'input[name="password"]'
  element :error_message, '#error'

  def submit_button
    Capybara.find_button('Submit')
  end
end

File: playwrite_js/page_objects/login_page.js
export class LoginPage {
   /**
   * @param {import('@playwright/test').Page} page
   */

    constructor( page ) {
        this.page = page;
    };

    goto() {
        return this.page.goto('practice-test-login');
    }
    
    user_name_field() {
        return this.page.getByLabel('Username');
    }

    password_field() {
        return this.page.getByLabel('Password');
    }

    submit_button() {
        return this.page.getByRole('button', { name: 'Submit' });
    }

    error_message() {
        return this.page.locator('div#error');
    }
}